src/connector/src/parser/unified/upsert.rs

```diff
@@ -86,34 +77,62 @@ where
             })
         }
     }
+
+    pub fn access_field_impl(&self, desc: &SourceColumnDesc) -> super::AccessResult {
+        match desc.additional_column.column_type {
+            Some(AdditionalColumnType::Key(_)) => self.access_key(&[&desc.name], &desc.data_type),
+            None => self.access_value(&[&desc.name], &desc.data_type),
+            _ => unreachable!(),
+        }
+    }
+}
+
+/// Wraps a key-value message into an upsert event, which uses `null` value to represent `DELETE`s.
+pub struct UpsertChangeEvent<K, V>(KvEvent<K, V>);
+
+impl<K, V> Default for UpsertChangeEvent<K, V> {
+    fn default() -> Self {
+        Self(KvEvent::default())
+    }
+}
+
+impl<K, V> Deref for UpsertChangeEvent<K, V> {
+    type Target = KvEvent<K, V>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<K, V> DerefMut for UpsertChangeEvent<K, V> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
 }
 
 impl<K, V> ChangeEvent for UpsertChangeEvent<K, V>
 where
     K: Access,
-    V: Access,
+    V: NullableAccess,
 {
     fn op(&self) -> std::result::Result<ChangeEventOperation, AccessError> {
-        if let Ok(Some(_)) = self.access_value(&[], None) {
-            Ok(ChangeEventOperation::Upsert)
+        if let Some(va) = &self.0.value_accessor {
+            if va.is_null() {
+                Ok(ChangeEventOperation::Delete)
+            } else {
+                Ok(ChangeEventOperation::Upsert)
+            }
         } else {
-            Ok(ChangeEventOperation::Delete)
+            Err(AccessError::Undefined {
+                name: "value".to_string(),
+                path: String::new(),
+            })
         }
     }
 
     fn access_field(&self, desc: &SourceColumnDesc) -> super::AccessResult {
-        match desc.additional_column.column_type {
-            Some(AdditionalColumnType::Key(_)) => {
-                if let Some(key_as_column_name) = &self.key_column_name
-                    && &desc.name == key_as_column_name
-                {
-                    self.access_key(&[], Some(&desc.data_type))
-                } else {
-                    self.access_key(&[&desc.name], Some(&desc.data_type))
-                }
-            }
-            None => self.access_value(&[&desc.name], Some(&desc.data_type)),
-            _ => unreachable!(),
-        }
+        self.0.access_field_impl(desc)
     }
 }
+
+pub type PlainEvent<K, V> = KvEvent<K, V>;
```

Comment:
Can we also leave some documentation and wrap it a new type? Otherwise the inlay hints from IDE can be confusing.

Reply:
hmm I think the `PlainEvent` is to represent an event consist of a key part and value part. And the name is ambiguous with the concept of `FORMAT PLAIN` and `PlainParser`, are they related?

Reply:
They are one thing. Basically plain means nothing